scarlet/wavelet.py

```diff
@@ -0,0 +1,295 @@
+        lvl = lvl_max
+    return lvl
+
+def mk_starlet(shape, image = None):
+    """ Creates a starlet for a given 2d shape.
+
+    Parameters
+    ----------
+    shape: tuple
+        2D shape of the desired shapelet
+    lvl: int
+        number of shapelet levels to compute. If None, lvl is set to the log2 of the number of pixels on a side.
+        if lvl is higher than this number lvl will be set to it.
+
+    Returns
+    -------
+    starlet: Fourier object
+        the starlet transform of a Dirac fonction as the `image` of a Fourier object
+
+    """
+    lvl, n1, n2 = shape[-3:]
+
+    # Filter size
+    n = np.size(h)
+    if image is None:
+        c = np.zeros((n1,n2))
+        c[int(n1/2), int(n2/2)] = 1
+    else:
+        if len(image.shape) > 2:
+            wave = []
+            for im in image:
+                wave.append(mk_starlet(shape, im))
+            return np.array(wave)
+        else:
+            c = image
+    c = fft.Fourier(c)
+    ## wavelet set of coefficients.
+    wave = np.zeros([lvl, n1, n2])
+    for i in np.arange(lvl - 1):
+        newh = np.zeros((n + (n - 1) * (2 ** i - 1), 1))
+        newh[0::2 ** i, 0] = h
+        newhT = fft.Fourier(newh.T)
+        newh = fft.Fourier(newh)
+
+        # Calculates c(j+1)
+        # Line convolution
+        cnew = fft.convolve(c, newh, axes=[0])
+
+        # Column convolution
+        cnew = fft.convolve(cnew, newhT, axes=[1])
+
+        ###### hoh for g; Column convolution
+        hc = fft.convolve(cnew, newh, axes=[0])
+
+        # hoh for g; Line convolution
+        hc = fft.convolve(hc, newhT, axes=[1])
+
+        # wj+1 = cj-hcj+1
+        wave[i, :, :] = c.image - hc.image
+
+        c = cnew
+
+    wave[-1, :, :] = c.image
+    return wave
+
+
+
+def iuwt(starlet):
+
+    """ Inverse starlet transform
+
+    Parameters
+    ----------
+    starlet: Shapelet object
+        Starlet to be inverted
+
+    Returns
+    -------
+    cJ: array
+        a 2D image that corresponds to the inverse transform of stralet.
+    """
+    lvl, n1, n2 = np.shape(starlet)
+    n = np.size(h)
+    # Coarse scale
+    cJ = fft.Fourier(starlet[-1, :, :])
+    for i in np.arange(1, lvl):
+        newh = np.zeros((n + (n - 1) * (2 ** (lvl - i - 1) - 1), 1))
+        newh[0::2 ** (lvl - i - 1), 0] = h
+        newhT = fft.Fourier(newh.T)
+        newh = fft.Fourier(newh)
+
+        # Line convolution
+        cnew = fft.convolve(cJ, newh, axes=[0])
+        # Column convolution
+        cnew = fft.convolve(cnew, newhT, axes=[1])
+
+        cJ = fft.Fourier(cnew.image + starlet[lvl - 1 - i, :, :])
+
+    return np.reshape(cJ.image, (n1, n2))
+
+
+class InputError(Exception):
+    """Exception raised for errors in the input.
+
+    Attributes:
+        expression -- input expression in which the error occurred
+        message -- explanation of the error
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+def mad_wavelet(image):
```

Comment:
This should be method of `Starlet`, then it could directly access its own coefficients. From what I can see, this is only used in this context anyway, so a free method is not needed.

Reply:
I disagree. This is a rather robust way of estimating the noise (assuming gaussianity) in a noise-dominated image, so it could easily be used on any images regardless of whether this image is to be modeled by wavelets.

Reply:
That makes no sense to me. The method is in a module called `wavelet` and internally computes the starlet coefficients (not some coefficients, starlet coefficients). This is redundant and not at all general. It should be a starlet method or property.

Reply:
Say I have an image for which I want to get the noise standard deviation. I would find it super awkward to have to do this: `sigma = Starlet(image).mad_wavelet` instead of `sigma = mad_wavelet(image)`. I'm fine with having this function in another module, but I don't see where else it could go.

Reply:
I strongly prefer `sigma = Starlet(image).mad`. It also avoids computing the starlet transform twice: once to get `noise` to call `StarletSource`, and the another time to set the starlet transform for that `StarletSource`.

Reply:
Nope, that wouldn't work like that, would it. Noise is computed from the transform of the image. The sources are set by the transform of the sources as initialised by `init_extended_source`, so it doesn't cut on anything. the Starlet(image).mad makes it look weird and obscure, when `sigma = mad_wavelet(image)` looks more straightforward and transparent: That's a function that gives you the mad.

Reply:
OK, fair point